src/HappyEyeBallsConnectionBuilder.php

```diff
@@ -0,0 +1,278 @@
+<?php
+
+namespace React\Socket;
+
+use React\Dns\Model\Message;
+use React\Dns\Resolver\ResolverInterface;
+use React\EventLoop\LoopInterface;
+use React\EventLoop\TimerInterface;
+use React\Promise;
+use React\Promise\CancellablePromiseInterface;
+
+/**
+ * @internal
+ */
+final class HappyEyeBallsConnectionBuilder
+{
+    const CHECK_INTERVAL = 0.1;
+
+    public $loop;
+    public $connector;
+    public $resolver;
+    public $uri;
+    public $host;
+    public $resolved = array(
+        Message::TYPE_A    => false,
+        Message::TYPE_AAAA => false,
+    );
+    public $resolverPromises = array();
+    public $connectionPromises = array();
+    public $connectQueue = array();
+    public $timer;
+    public $parts;
+    public $ipsCount = 0;
+    public $failureCount = 0;
+    public $resolve;
+    public $reject;
+
+    public function __construct(LoopInterface $loop, ConnectorInterface $connector, ResolverInterface $resolver, $uri, $host, $parts)
+    {
+        $this->loop = $loop;
+        $this->connector = $connector;
+        $this->resolver = $resolver;
+        $this->uri = $uri;
+        $this->host = $host;
+        $this->parts = $parts;
+    }
+
+    public function connect()
+    {
+        $that = $this;
+        return new Promise\Promise(function ($resolve, $reject) use ($that) {
+            /**
+             * Delay A lookup by 50ms as per RFC.
+             *
+             * @link https://tools.ietf.org/html/rfc8305#section-3
+             */
+            $that->resolverPromises[Message::TYPE_AAAA] = $that->resolve(Message::TYPE_AAAA, $resolve, $reject);
+            $that->resolverPromises[Message::TYPE_A] = Promise\Timer\resolve(0.05, $that->loop)->then(function () use ($that, $resolve, $reject) {
+                $that->resolverPromises[Message::TYPE_A] = $that->resolve(Message::TYPE_A, $resolve, $reject);
+            });
```

Comment:
https://tools.ietf.org/html/rfc8305#section-3 appears to be a bit unclear, but to me it suggests both queries should be started simultaneously. If `AAAA` resolves first, connection attempts can start immediately. If `A` resolves first, it should wait 50ms for the `AAAA` response before starting connection attempts.

This appears to differ from the current implementation which always waits for 50ms to start the `A` query no matter what.

Reply:
Correct, my intention was to add that later on. But I suppose we can just as easily add it now

Reply:
IMHO I read that as send them both out as fast as possible after each other.

Reply:
Implemented